New feature: Save Lab #1587

junkiattan opened the pull request:
#1587 

There are two functionalities for Save Lab.

* The first is the actual save functionality. Users can use it to save selected slides as a new presentation, preserving notes, animations, images, etc.

* Another is the setting of the default save location. By default the save location would be in the user's Documents folder. However this can be set in the Save Lab settings to any other folder and the user choice will be remembered every subsequent time that he opens powerpoint.

I have also added a FT for SaveLab. In order to run it properly, please follow the [steps](https://www.ispringsolutions.com/blog/installing-vba-component/) to enable VBA component in your powerpoint, or else the default Presentations method would fail to work.

I have tested the code and FT and its works for PPT2016 and PPT2013.

Some screenshots of the new feature

<img width="960" alt="savelab1" src="https://user-images.githubusercontent.com/16800272/37603718-2f156ee0-2bca-11e8-9662-4d395b8107c8.png">
<img width="960" alt="savelab2" src="https://user-images.githubusercontent.com/16800272/37603720-302c62d4-2bca-11e8-9a16-e987fbe4cd69.png">
<img width="960" alt="savelab3" src="https://user-images.githubusercontent.com/16800272/37603722-30a5cbf6-2bca-11e8-8cad-959549e26b5c.png">

leeyh20 commented:
I originally got this error while testing:
System.NullReferenceException
Object reference not set to an instance of an object.
Stack Trace:
   at System.Drawing.Bitmap..ctor(Image original)
   at PowerPointLabs.ActionFramework.SaveLab.SaveLabMenuImageHandler.GetImage(String ribbonId) in 
Then I realised I just needed to **rebuild** the entire solution to refresh the Resources.

leeyh20 commented on PowerPointLabs/PowerPointLabs/TextCollection/SaveLabText.cs:
```diff
@@ -0,0 +1,28 @@
+
+        #region GUI Text
+        public const string RibbonMenuLabel = "Save";
+        public const string SavePresentationsButtonLabel = "Save Presentations";
```
Should this be "Save slides" or "Export slides" instead?

junkiattan replied:
Lets remain as "Save Presentation" for greater clarity as it belongs to Save Lab

leeyh20 replied:
Ok then maybe shouldn't be plural form. Yeah "Save Presentation" sounds okay...

junkiattan commented on PowerPointLabs/PowerPointLabs/SaveLab/SaveLabMain.cs:
```diff
@@ -0,0 +1,69 @@
+                try
+                {
+                    // Re-open the save copy in the same directory in the background
+                    Presentations newPres = new Microsoft.Office.Interop.PowerPoint.Application().Presentations;
```
Ultimately it would call base.Open(), which would still invoke the Presentations.Open() method, which would throw an error unless you install the VBA component. 

Note that the open methods you are referring to are in the main project, not in the test project. The VBA component is for the testing, and does not affect the main functionality. Using Open method in Test project is much more trickier, which needs the workaround of the VBA component.

junkiattan commented on PowerPointLabs/PowerPointLabs/SaveLab/SaveLabMain.cs:
```diff
@@ -0,0 +1,69 @@
+                    for (int i = newPresentation.SlideCount - 1; i >= 0; i--)
+                    {
+                        removeSlide = true;
+                        foreach (Models.PowerPointSlide selectedSlide in selectedSlides)
```
Thanx for the optimization suggestion. Ive implemented it!

junkiattan commented on PowerPointLabs/PowerPointLabs/SaveLab/SaveLabSettings.cs:
```diff
@@ -0,0 +1,89 @@
+            }
+        }
+
+        // Function updates the stpred path in the local storage with a new path
```
Rectified

junkiattan commented:
Changes made:
1. Implemented Hashset for more efficient searching
1. Added functionality to check for and remove empty sections
1. Changed the settings icon for consistency

@leeyh20 Awaiting approval

junkiattan commented:
@initialshl Added the required pptx to store icons for Save Lab and resolved merge conflicts

initialshl commented:
@junkiattan Theres another merge conflict, can merge once its resolved
